library/Fission/Web/Auth/JWT.hs

```diff
@@ -0,0 +1,124 @@
+
+handler ::
+  ( MonadIO     m
+  , MonadLogger m
+  , MonadThrow  m
+  , MonadDB     m
+  )
+  => Auth.BearerToken
+  -> m (Entity User)
+handler token = 
+  validateJWT token >>= \case
+    Left err -> do
+      logWarn <| "Failed login with token " <> Auth.unBearer token
+      throwM err
+    Right did -> do
+      mayUser <- runDB <| selectFirst
+        [ UserDid    P.==. decodeUtf8Lenient did
+        , UserActive P.==. True
+        ] []
+      
+      case mayUser of
+        Nothing -> throwM JWT.NoUser
+        Just usr -> return usr
+
+validateJWT :: MonadIO m => Auth.BearerToken -> m (Either JWT.Error ByteString)
+validateJWT token = 
+  case parseJWT token of
+    Left err -> return <| Left err
+    Right pl ->
+      validateTime pl >>= \case
+        Left err -> return <| Left err
+        Right _ -> return <| Right <| iss pl
+
+parseJWT :: Auth.BearerToken -> Either JWT.Error JWT.Payload
+parseJWT token = do
+  (headerRaw, payloadRaw, sig64) <- getParts <| Auth.unBearer token
+  _ <- validateHeader headerRaw
+  payload <- decodePart payloadRaw
+  let 
+    content = headerRaw <> "." <> payloadRaw
+    did = iss payload
+  pubkey64 <- didToPubkey did
+  _ <- validateSig content pubkey64 sig64
+  Right payload
+
+getParts :: ByteString -> Either JWT.Error (ByteString, ByteString, ByteString)
+getParts token = do
+  let parts = Ch.split '.' token
+  case parts of
+    [header, payload, sig] -> Right (header, payload, sig)
+    _ -> Left ParseError
+
+validateHeader :: ByteString -> Either JWT.Error ()
+validateHeader bytes = 
+  case decodePart bytes of
+    Left err -> Left err
+    Right header -> 
+      if typ header /= "JWT"
+      then Left BadHeader
+      else
+        if alg header /= "Ed25519"
+        then Left UnsupportedAlg
+        else Right ()
+
+validateTime :: MonadIO m => JWT.Payload -> m (Either JWT.Error ())
+validateTime pl = do
+  time <- getCurrentPOSIXTime
+  if time > JWT.exp pl
+  then return <| Left JWT.Expired
+  else
+    if time < JWT.nbf pl
+    then return <| Left JWT.TooEarly
+    else return <| Right ()
+
+didToPubkey :: ByteString -> Either JWT.Error ByteString
+didToPubkey did = 
+  case BS.stripPrefix "did:nacl:" did of
+    Nothing -> Left JWT.DIDNotSupported
+    Just pubkey -> Right pubkey
+
+validateSig :: ByteString -> ByteString -> ByteString -> Either JWT.Error ()
+validateSig content pubkey64 sig64 = 
+  case Crypto.base64ToEdPubKey pubkey64 of
+    CryptoFailed _ -> Left BadPublicKey
+    CryptoPassed pk ->
+      case Crypto.base64ToSignature sig64 of
+        CryptoFailed _ -> Left BadSignature
+        CryptoPassed sig -> do
+          case Ed.verify pk (Crypto.pack content) sig of
+            False -> Left IncorrectSignature
+            True -> Right ()
+
+decodePart :: FromJSON a => ByteString -> Either JWT.Error a
+decodePart bytes = bytes 
+  |> Base64.decodeLenient
+  |> BS.Lazy.fromStrict
+  |> decode
+  |> maybe (Left ParseError) Right
```

Comment:
Really nit-picky and optional, but as we use more and more of pipes, I think that this is the most consistent:

```haskell
decodePart bytes = 
  bytes 
    |> Base64.decodeLenient
    |> BS.Lazy.fromStrict
    |> decode
    |> maybe (Left ParseError) Right
```